docs/tech-notes/life_of_a_query.md

```diff
@@ -825,214 +881,264 @@ txn). This code realizes that, if a request from the current txn has
+and [passes execution](https://github.com/cockroachdb/cockroach/blob/759fdac56ec59c75aa872e8e00d0a614bcc593a3/pkg/kv/kvserver/replica_evaluate.go#L506-L510)
+to it.
+
+One typical read request is a [`ScanRequest`](https://github.com/cockroachdb/cockroach/blob/4e4f31a0ed1a8ea985b9ab6f72e29266b259900e/pkg/roachpb/api.proto#L384),
+which is evaluated by
+[`batcheval.Scan`](https://github.com/cockroachdb/cockroach/blob/261fc356765e61215478ba697d4fe252b9466c8f/pkg/kv/kvserver/batcheval/cmd_scan.go#L31).
+The code is rather brief, and mostly
+[calls](https://github.com/cockroachdb/cockroach/blob/261fc356765e61215478ba697d4fe252b9466c8f/pkg/kv/kvserver/batcheval/cmd_scan.go#L54-L55)
+a corresponding function for the underlying storage engine.
+
+#### Engine
+
+We're getting to the bottom of the CockroachDB stack. The
+[`Engine`](https://github.com/cockroachdb/cockroach/blob/83185789481f23bd6c4938a3588203f4cd4eef89/pkg/storage/engine.go#L640)
+is an interface abstracting away different on-disk stores. From
+version 21.1 onwards, we only support [Pebble](https://github.com/cockroachdb/pebble),
+an embedded key/value database we developed ourselves based on
+[RocksDB](https://rocksdb.org)/[LevelDB](https://github.com/google/leveldb).
+RocksDB was previous used as the storage engine, but was replaced in pursuit of
+[performance, stability, and tighter integration](https://www.cockroachlabs.com/blog/pebble-rocksdb-kv-store/).
+Even though the storage engine is a crucial part of servicing a request, we
+won't go into any details on Pebble's internals here, in the interest of
+brevity.
+
+For reads, we typically make use of the
+[`storage.Reader`](https://github.com/cockroachdb/cockroach/blob/83185789481f23bd6c4938a3588203f4cd4eef89/pkg/storage/engine.go#L349)
+interface, a subset of `Engine`. The `Scan` command from the previous
+section passes this `Reader` to [`storage.MVCCScanToBytes()`](https://github.com/cockroachdb/cockroach/blob/129b2dcaecfce350d4ed3a4b1f6b4b654b0bd11e/pkg/storage/mvcc.go#L2585),
+which is a helper function that fetches a list of key/value pairs in a
+given key span from the storage engine.
+
+The CockroachDB storage layer uses [Multi-Version Concurrency Control (MVCC)](https://en.wikipedia.org/wiki/Multiversion_concurrency_control),
+storing multiple versions of Protobuf-encoded [`Value`s](https://github.com/cockroachdb/cockroach/blob/0254e86051c310d38e5668cf82ba128ff4c1ab28/pkg/roachpb/data.proto#L84)
+for a given key. Each version is identified by the value's
+[`Timestamp`](https://github.com/cockroachdb/cockroach/blob/0254e86051c310d38e5668cf82ba128ff4c1ab28/pkg/roachpb/data.proto#L90),
+and the latest version visible to a given transaction is given by its
+[read timestamp](https://github.com/cockroachdb/cockroach/blob/0254e86051c310d38e5668cf82ba128ff4c1ab28/pkg/roachpb/data.proto#L388).
+As a special case, new values (called write intents) are written
+without a timestamp, and once the transaction commits these will
```

Comment:
depending on the point of view that's right or not quite right (intents contain a timestamp, which is the lowest timestamp the intent could ever commit at, but they are written as inline values into the meta key which is not a timestamped key). Seems fine, perhaps even wise, to leave as is.

Reply:
Yeah, in this context we take "timestamp" to mean "version", and the definition of intents in MVCC is values with no timestamp/version.